File: fixtures/rust-sample/src/inheritance.rs
// cgx-fixture: inheritance and override chains
// Covers: impl Trait for T (implements edges), overrides of default methods,
//         super-trait calls, struct embedding pattern

pub trait Animal {
    fn name(&self) -> &str;

    // Default implementation — can be overridden
    fn description(&self) -> String {
        format!("{} is an animal", self.name())
    }

    fn sound(&self) -> String;
}

pub trait Trainable: Animal {
    fn train(&self, command: &str) -> bool;

    // Default: calls sound() from Animal
    fn respond(&self) -> String {
        format!("{} responds: {}", self.name(), self.sound())
    }
}

pub struct Labrador {
    pub name: String,
}

pub struct Poodle {
    pub name: String,
}

impl Animal for Labrador {
    fn name(&self) -> &str {
        &self.name
    }

    fn sound(&self) -> String {
        "woof".to_string()
    }
    // description() uses default implementation — calls self.name() via default
}

impl Animal for Poodle {
    fn name(&self) -> &str {
        &self.name
    }

    fn sound(&self) -> String {
        "yip".to_string()
    }

    // Override: does not call the default
    fn description(&self) -> String {
        format!("{} is a fancy poodle", self.name())
    }
}

impl Trainable for Labrador {
    fn train(&self, command: &str) -> bool {
        !command.is_empty()
    }
    // respond() uses default — calls self.name() and self.sound() via default
}

impl Trainable for Poodle {
    fn train(&self, command: &str) -> bool {
        command.len() < 10
    }

    // Overrides respond — custom implementation
    fn respond(&self) -> String {
        format!("{} does a trick!", self.name())
    }
}

/// Demonstrate virtual dispatch to overridden methods.
pub fn describe_animal(a: &dyn Animal) -> String {
    a.description()   // calls:virtual -> [Labrador::description(default), Poodle::description(override)]
}

pub fn get_response(t: &dyn Trainable) -> String {
    t.respond()       // calls:virtual -> [Labrador::respond(default), Poodle::respond(override)]
}

/// Check override chain: default calls through to overridden method.
pub fn default_description_chain() -> String {
    let lab = Labrador { name: "Rex".to_string() };
    lab.description()  // calls default, which calls self.name() (certain, Labrador::name)
}

File: fixtures/ts-sample/src/inheritance.ts
// cgx-fixture: inheritance and override chains (TypeScript)
// Covers: extends (inherits edge), implements (implements edge),
//         method overrides (overrides edge), super calls, abstract methods

export interface Animal {
    name(): string;
    sound(): string;
    description?(): string;
}

export abstract class BaseAnimal implements Animal {
    private _name: string;

    constructor(name: string) {
        this._name = name;
    }

    name(): string {
        return this._name;  // concrete method, no override
    }

    abstract sound(): string;  // must be overridden

    // Default description — calls this.name() and this.sound() (virtual)
    description(): string {
        return `${this.name()} says ${this.sound()}`;  // calls:virtual on sound()
    }
}

export class Labrador extends BaseAnimal {
    constructor(name: string) {
        super(name);   // super call — calls BaseAnimal constructor
    }

    sound(): string {
        return "woof";  // overrides abstract method
    }
    // description() inherited — uses BaseAnimal::description
}

export class Poodle extends BaseAnimal {
    constructor(name: string) {
        super(name);
    }

    sound(): string {
        return "yip";
    }

    // Override: custom description, calls super.description()
    description(): string {
        const base = super.description();   // super call: certain -> BaseAnimal::description
        return `${base} (fancy)`;
    }
}

export interface Trainable {
    train(command: string): boolean;
    respond(): string;
}

export class TrainedLabrador extends Labrador implements Trainable {
    constructor(name: string) {
        super(name);
    }

    train(command: string): boolean {
        return command.length > 0;
    }

    respond(): string {
        return `${this.name()} sits!`;  // certain: this.name() on concrete class
    }
}

/// Virtual dispatch to overridden method.
export function describeAnimal(a: Animal): string {
    return a.description ? a.description() : a.sound();  // calls:virtual, conditional
}

/// Super chain: Poodle.description -> BaseAnimal.description.
export function poodleDescription(): string {
    const p = new Poodle("Fifi");
    return p.description();  // calls Poodle::description which super-calls BaseAnimal::description
}

/// Array of base type — virtual dispatch on each element.
export function allDescriptions(animals: BaseAnimal[]): string[] {
    return animals.map(a => a.description());  // calls:virtual inside map callback
}
